components/org.wso2.carbon.identity.api.user.recovery/org.wso2.carbon.identity.rest.api.user.recovery.v2/src/main/java/org/wso2/carbon/identity/rest/api/user/recovery/v2/impl/core/utils/RecoveryUtil.java

```diff
@@ -0,0 +1,517 @@
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.impl.core.exceptions.ForbiddenException;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.impl.core.exceptions.NotAcceptableException;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.impl.core.exceptions.NotFoundException;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.impl.core.exceptions.PreconditionFailedException;
+
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.APICall;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.Property;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.RecoveryChannel;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.RetryErrorResponse;
+import org.wso2.carbon.identity.rest.api.user.recovery.v2.model.UserClaim;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.Response;
+
+import static org.wso2.carbon.identity.api.user.common.Constants.TENANT_CONTEXT_PATH_COMPONENT;
+import static org.wso2.carbon.identity.application.common.util.IdentityApplicationConstants.Error.UNEXPECTED_SERVER_ERROR;
+
+/**
+ * Contains the recovery endpoint utils.
+ */
+public class RecoveryUtil {
+
+    private static final Log log = LogFactory.getLog(RecoveryUtil.class);
+    private static final String LOG_MESSAGE_PREFIX = "INITIATOR";
+    private static final String FORBIDDEN_ERROR_CATEGORY = "FORBIDDEN_ERROR_CATEGORY";
+    private static final String CONFLICT_REQUEST_ERROR_CATEGORY = "CONFLICT_REQUEST_ERROR_CATEGORY";
+    private static final String REQUEST_NOT_FOUND_ERROR_CATEGORY = "REQUEST_NOT_FOUND_ERROR_CATEGORY";
+    private static final String REQUEST_NOT_ACCEPTABLE_ERROR_CATEGORY = "REQUEST_NOT_ACCEPTABLE_ERROR_CATEGORY";
+    private static final String RETRY_ERROR_CATEGORY = "RETRY_ERROR_CATEGORY";
+
+    // Map with the error codes categorized in to different error groups.
+    private static final HashMap<String, String> clientErrorMap = generateClientErrorMap();
+
+    /**
+     * Converts a list of UserClaim in to a UserClaim array.
+     *
+     * @param userClaimsList UserClaims List.
+     * @return Map of user claims.
+     */
+    public static HashMap<String, String> buildUserClaimsMap(List<UserClaim> userClaimsList) {
+
+        HashMap<String, String> userClaims = new HashMap<>();
+        for (UserClaim userClaimModel : userClaimsList) {
+            userClaims.put(userClaimModel.getUri(), userClaimModel.getValue());
+        }
+        return userClaims;
+    }
+
+    /**
+     * Convert the list of Properties in to an array.
+     *
+     * @param propertyList List of {@link Property} objects.
+     * @return Map of properties.
+     */
+    public static HashMap<String, String> buildPropertiesMap(List<Property> propertyList) {
+
+        HashMap<String, String> properties = new HashMap<>();
+        if (propertyList == null) {
```

Comment:
why would this be null? if this is passed by your code, make sure you pass empty list instead of null

Reply:
Since `properties` is not a required attribute in the request body, when properties are not sent, `propertyList` will be set as null.